### DISCUSSION TITLE
Improve security for username/token and allow compatibility with Nexus Repository Manager

### DISCUSSION THREAD

**Discussion by lfvjimisola** [npm/feedback#776] [CLI] [answered]
Nexus Repository Manager uses mixed case for username and token (for improved security I presume) but npm does only allow lowercase:

```
npm WARN Name must be lowercase 
Username: (Ey3LtLmE) 
```

Sonatype are reluctant to changing their behaviour. Is there any change that you could improve?

Related: https://issues.sonatype.org/browse/NEXUS-19920

**Comment by ljharb** [comment]
If something trying to be an npm registry doesn't match the constraints an npm registry client follows, then it seems like it's broken?

**Comment by MylesBorins** [comment, marked as answer]
We don't allow mixed case user names as at one time the registry was case sensitive and there were instances of duplicated usernames. The solution that we landed on was to limit the registry to only lowercase characters for usernames and to deprecate all mixed case names to avoid having more than 1 username where the only difference was casing... this was and remains a security risk if we were to change this behavior.

Does sonatype support any sort of access token? If you were to generate the access token in the sonatype UI you could utilize that and avoid the whole base64 encoded basic auth approach they suggest.

The npm public registry, and the majority of registry products on the market, do not use randomly generated user names, and as such we are unlikely to support this feature. Unless the sonatype registry is case sensitive, which opens up all the problems we mentioned above, there shouldn't be a reason to avoid lowercasing at least the API they use for authenticating. If user names are case sensitive there is an even bigger divergence from the public registry that i think would be worth bringing up to them... as this does not "improve security".

If you are just trying to unblock you specific workflow you could likely monkey patch [and comment out these lines](https://github.com/npm/cli/blob/latest/node_modules/npm-user-validate/npm-user-validate.js#L24-L26) in your local CLI to avoid the lowercase check, although I don't advise it.

**Reply by brianf** [reply]
It seems like the issue is, if I understand it correctly, the user tokens that are generated are now incompatible with what the npm client allows. Is that a correct interpretation? 

I'm trying to tease out, is it the npm client doesn't allow mixed case anymore or is it just the registry? (if it's the later then I'm not understanding the problem)

**Reply by MylesBorins** [reply]
neither the client or registry support mixed case. To be clear this is for usernames and package names only... not authentication tokens. Our access tokens for the registry are similarly base64 encoded so there is no issue with mixed cases in the access token... the issue is only with the username having mixed casing
